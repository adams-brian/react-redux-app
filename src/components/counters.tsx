import * as React from 'react';
import { Counter } from './counter';

interface ICountersProps {
  counters: number[];
  addCounter: () => void;
  increment: (index: number) => void;
  decrement: (index: number) => void;
  reset: (index: number) => void;
  remove: (index: number) => void;
}

/* tslint:disable:jsx-no-lambda */
export const Counters = (props: ICountersProps) => (
  <div className="counters-container">
    <h1>Counters</h1>
    <div className="counter-container">
      {props.counters.map((counter, index) => 
        (<Counter 
          key={index} /* not ideal, but it works in this case */
          counter={counter}
          increment={() => props.increment(index)}
          decrement={() => props.decrement(index)}
          reset={() => props.reset(index)}
          remove={() => props.remove(index)}
        />))}
      <div className="add-counter rounded" onClick={props.addCounter}>
        <span className="fa fa-plus" />
      </div>
    </div>
  </div>
);
