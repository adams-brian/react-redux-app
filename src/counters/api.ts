export const saveCounters = (counters: number[]) =>
  fetch('http://localhost:4000/counters', {
    body: JSON.stringify({counters}),
    headers: {
      'Accept': 'application/json, text/plain, */*',
      'Content-Type': 'application/json'
    },
    method: 'post',
  })

export const loadCounters = async (): Promise<number[]> => {
  const response = await fetch('http://localhost:4000/counters');
  const json = await response.json();
  return [...json.data];
}
