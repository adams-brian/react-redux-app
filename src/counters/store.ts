import { IAction } from '../store';

export interface IState {
  counters: number[];
}

/* tslint:disable:max-classes-per-file */
abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
export type CountersAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const actionCreators = {
  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),
};

function countersReducer(state: number[] = [], action: CountersAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

export const reducers = {
  counters: countersReducer
}
