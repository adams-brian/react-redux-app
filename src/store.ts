import { LocationChangeAction } from 'react-router-redux';

export interface IUser { 
  _id: string;
  firstname: string;
  lastname: string;
}

export interface IAppState {
  loading: boolean;
  counters: number[];
  users: IUser[];
}

interface IAction {
  type: string;
}

/* tslint:disable:max-classes-per-file */
export const START_LOADING = '[App] START LOADING';
class StartLoading implements IAction {
  public readonly type = START_LOADING;
}
export const DONE_LOADING = '[App] DONE LOADING';
class DoneLoading implements IAction {
  public readonly type = DONE_LOADING;
}
type LoadingAction = StartLoading | DoneLoading;

abstract class ActionWithNumberPayload implements IAction {
  public readonly type: string;
  constructor(public payload: number) {}
}
export const INCREMENT = '[Counters] INCREMENT';
class Increment extends ActionWithNumberPayload {
  public readonly type = INCREMENT;
}
export const DECREMENT = '[Counters] DECREMENT';
class Decrement extends ActionWithNumberPayload {
  public readonly type = DECREMENT;
}
export const RESET = '[Counters] RESET';
class Reset extends ActionWithNumberPayload {
  public readonly type = RESET;
}
export const ADD_COUNTER = '[Counters] ADD COUNTER';
class AddCounter implements IAction {
  public readonly type = ADD_COUNTER;
}
export const REMOVE_COUNTER = '[Counters] REMOVE COUNTER';
class RemoveCounter extends ActionWithNumberPayload {
  public readonly type = REMOVE_COUNTER;
}
export const COUNTERS_UPDATED = '[Counters] COUNTERS UPDATED';
class CountersUpdated implements IAction {
  public readonly type = COUNTERS_UPDATED;
  constructor(public payload: number[]) {}
}
type CounterAction = Increment | Decrement | Reset | AddCounter | RemoveCounter | CountersUpdated;

export const CREATE_USER = '[Users] CREATE';
export class CreateUser implements IAction {
  public readonly type = CREATE_USER;
  constructor(public payload: IUser) {}
}
export const UPDATE_USER = '[Users] UPDATE';
class UpdateUser implements IAction {
  public readonly type = UPDATE_USER;
  constructor(public payload: IUser) {}
}
export const DELETE_USER = '[Users] DELETE';
export class DeleteUser implements IAction {
  public readonly type = DELETE_USER;
  constructor(public payload: string) {}
}
export const USERS_UPDATED = '[Users] USERS UPDATED';
class UsersUpdated implements IAction {
  public readonly type = USERS_UPDATED;
  constructor(public payload: IUser[]) {}
}
type UserAction = CreateUser | UpdateUser | DeleteUser | UsersUpdated;

export const actionCreators = {
  doneLoading: () => Object.assign({}, new DoneLoading()),
  startLoading: () => Object.assign({}, new StartLoading()),

  addCounter: () => Object.assign({}, new AddCounter()),
  countersUpdated: (counters: number[]) => Object.assign({}, new CountersUpdated(counters)),
  decrement: (index: number) => Object.assign({}, new Decrement(index)),
  increment: (index: number) => Object.assign({}, new Increment(index)),
  removeCounter: (index: number) => Object.assign({}, new RemoveCounter(index)),
  reset: (index: number) => Object.assign({}, new Reset(index)),

  createUser: (user: IUser) => Object.assign({}, new CreateUser(user)),
  deleteUser: (id: string) => Object.assign({}, new DeleteUser(id)),
  updateUser: (user: IUser) => Object.assign({}, new UpdateUser(user)),
  usersUpdated: (users: IUser[]) => Object.assign({}, new UsersUpdated(users)),
};

function loadingReducer(state: boolean = false, action: LoadingAction) {
  switch (action.type) {
    case START_LOADING:
      return true;
    case DONE_LOADING:
      return false;
    default:
      return state;
  }
}

function countersReducer(state: number[] = [], action: CounterAction) {
  switch (action.type) {
    case INCREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] + 1,
        ...state.slice(action.payload + 1)
      ];
    case DECREMENT:
      return [
        ...state.slice(0, action.payload),
        state[action.payload] - 1,
        ...state.slice(action.payload + 1)
      ];
    case RESET:
      return [
        ...state.slice(0, action.payload),
        0,
        ...state.slice(action.payload + 1)
      ];
    case ADD_COUNTER:
      return [...state, 0];
    case REMOVE_COUNTER:
      return [
        ...state.slice(0, action.payload),
        ...state.slice(action.payload + 1)
      ];
    case COUNTERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
}

const usersReducer = (state: IUser[] = [], action: UserAction) => {
  switch (action.type) {
    case CREATE_USER:
      return [...state, action.payload];
    case UPDATE_USER:
      return state.map(user => user._id === action.payload._id ? Object.assign({}, action.payload) : user);
    case DELETE_USER:
      return state.filter(user => user._id !== action.payload);
    case USERS_UPDATED:
      return [...action.payload];
    default:
      return state;
  }
};

export const reducers = {
  counters: countersReducer,
  loading: loadingReducer,
  users: usersReducer
};

export type AppAction = LoadingAction | CounterAction | UserAction | LocationChangeAction;
