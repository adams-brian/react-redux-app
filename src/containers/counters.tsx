import { connect } from 'react-redux';
import { AppState, actionCreators } from '../store';

import { Counters as UnwrappedCounters } from '../components/counters';

export const Counters = connect(
  (state: AppState) => 
    ({ counters: state.counters }),
  { 
    addCounter: actionCreators.addCounter,
    increment: actionCreators.increment,
    decrement: actionCreators.decrement,
    reset: actionCreators.reset,
    remove: actionCreators.removeCounter
  }
  )(UnwrappedCounters);
  