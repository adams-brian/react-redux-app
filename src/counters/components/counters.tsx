import * as React from 'react';
import { connect } from 'react-redux';

import { actionCreators, IAppState } from '../../store';
import Counter from './counter';

interface ICountersProps {
  counters: number[];
  addCounter: () => void;
  increment: (index: number) => void;
  decrement: (index: number) => void;
  reset: (index: number) => void;
  remove: (index: number) => void;
}

/* tslint:disable:jsx-no-lambda */
export const Counters = (props: ICountersProps) => (
  <div className="counters-container">
    <h1>Counters</h1>
    <div className="counter-container">
      {props.counters.map((counter, index) => 
        (<Counter 
          key={index} /* not ideal, but it works in this case */
          counter={counter}
          increment={() => props.increment(index)}
          decrement={() => props.decrement(index)}
          reset={() => props.reset(index)}
          remove={() => props.remove(index)}
        />))}
      <div className="add-counter rounded" onClick={props.addCounter}>
        <span className="fa fa-plus" />
      </div>
    </div>
  </div>
);

export default connect(
  (state: IAppState) => 
    ({ counters: state.counters }),
  { 
    addCounter: actionCreators.addCounter,
    decrement: actionCreators.decrement,
    increment: actionCreators.increment,
    remove: actionCreators.removeCounter,
    reset: actionCreators.reset
  }
)(Counters);
  
