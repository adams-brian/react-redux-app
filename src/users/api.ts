import { IUser } from './store';

export const deleteUser = (id: string) =>
  fetch('http://localhost:4000/users/' + id, {
    headers: {
      'Accept': 'application/json, text/plain, */*',
      'Content-Type': 'application/json'
    },
    method: 'delete'
  })

export const loadUsers = async () => {
  const response = await fetch('http://localhost:4000/users');
  const json = await response.json();
  return [...json.data];
}

export const createUser = (user: IUser) =>
  fetch('http://localhost:4000/users', {
    body: JSON.stringify(user),
    headers: {
      'Accept': 'application/json, text/plain, */*',
      'Content-Type': 'application/json'
    },
    method: 'put'
  })

export const updateUser = (user: IUser) =>
  fetch('http://localhost:4000/users/' + user._id, {
    body: JSON.stringify(user),
    headers: {
      'Accept': 'application/json, text/plain, */*',
      'Content-Type': 'application/json'
    },
    method: 'post'
  })
