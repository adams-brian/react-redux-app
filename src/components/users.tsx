import * as React from 'react';
import { Route, Switch } from 'react-router-dom';

import { EditUser } from '../containers/editUser';
import { UserList } from '../containers/userList';

export const Users = () => (
  <Switch>
    <Route exact={true} path="/users" component={UserList}/>
    <Route exact={true} path="/users/createuser" component={EditUser}/>
    <Route path="/users/:id" component={EditUser}/>
  </Switch>
);
