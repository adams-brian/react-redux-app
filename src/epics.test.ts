import { Action } from 'redux';
import { ActionsObservable, StateObservable } from 'redux-observable';
import { Subject } from 'rxjs';
import * as sinon from 'sinon';

import * as countersApi from './counters/api';
import { actionCreators as countersActions } from './counters/store';
import { createUserEpic, deleteUserEpic, loadCountersEpic,
  loadUsersEpic, saveCountersEpic, updateUserEpic } from './epics';
import * as usersApi from './users/api';
import { actionCreators as usersActions } from './users/store';

describe('epics', () => {

  let actionSubject: Subject<Action>;
  let action$: ActionsObservable<Action>;
  let stateSubject: Subject<any>;
  let state$: StateObservable<any>;
  let output: Action[];

  beforeEach(() => {
    actionSubject = new Subject<Action>();
    action$ = ActionsObservable.from(actionSubject);
    stateSubject = new Subject<any>();
    state$ = new StateObservable(stateSubject, {});
    output = [];
  })

  describe('loadCountersEpics', () => {

    let loadCountersStub: sinon.SinonStub;
    const createLoadCountersStub = (fake: () => void) => 
      sinon.stub(countersApi, 'loadCounters').callsFake(fake)

    beforeEach(() => {
      loadCountersStub = createLoadCountersStub(
        () => Promise.resolve([5,1,4,2,3])
      );
      loadCountersEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      loadCountersStub.restore();
    });

    it('should dispatch CountersUpdated and CountersLoaded on LoadCounters if loadCounters succeeds', () => {
      actionSubject.next(countersActions.loadCounters());
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(loadCountersStub.calledOnce).toBe(true);
      });
    });

    it('should dispatch CountersError if loadCounters rejects', () => {
      // replace the standard stub with one that rejects
      loadCountersStub.restore();
      loadCountersStub = createLoadCountersStub(
        () => Promise.reject('error in loadCounters')
      );
      actionSubject.next(countersActions.loadCounters());
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(loadCountersStub.calledOnce).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(countersActions.addCounter());
      actionSubject.next(countersActions.removeCounter(1));
      actionSubject.next(countersActions.countersUpdated([3,2,1]));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(loadCountersStub.called).toBe(false);
      });
    });

  });

  describe('saveCountersEpic', () => {

    let clock: sinon.SinonFakeTimers;
    const list = [3,1,2];
    let saveCountersStub: sinon.SinonStub;
    const createSaveCountersStub = (fake: () => void) => 
      sinon.stub(countersApi, 'saveCounters').callsFake(fake)
    
    const testSaveCalled = (action: Action<any>) => {
      actionSubject.next(action);
      return Promise.resolve()
      .then(() => {
        expect(output).toEqual([]);
        expect(saveCountersStub.called).toBe(false);
        clock.tick(1000);
      })
      .then(() => {
        expect(output).toEqual([]);
        expect(saveCountersStub.calledOnceWithExactly(list)).toBe(true);
      });
    }

    beforeEach(() => {
      clock = sinon.useFakeTimers();
      saveCountersStub = createSaveCountersStub(
        () => Promise.resolve()
      );
      stateSubject.next({ counters: { list } });
      saveCountersEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      saveCountersStub.restore();
      clock.restore();
    });

    it('should call saveCounters on Increment', () => {
      testSaveCalled(countersActions.increment(1));
    });

    it('should call saveCounters on Decrement', () => {
      testSaveCalled(countersActions.decrement(1));
    });

    it('should call saveCounters on Reset', () => {
      testSaveCalled(countersActions.reset(1));
    });

    it('should call saveCounters on AddCounter', () => {
      testSaveCalled(countersActions.addCounter());
    });

    it('should call saveCounters on RemoveCounter', () => {
      testSaveCalled(countersActions.removeCounter(1));
    });

    it('should call saveCounters on a 1 second debounce', () => {
      actionSubject.next(countersActions.increment(1));
      actionSubject.next(countersActions.increment(1));
      return Promise.resolve()
      .then(() => {
        expect(saveCountersStub.called).toBe(false);
        clock.tick(900);
      })
      .then(() => {
        actionSubject.next(countersActions.decrement(0));
        actionSubject.next(countersActions.removeCounter(2));
        expect(saveCountersStub.called).toBe(false);
        clock.tick(900);
      })
      .then(() => {
        actionSubject.next(countersActions.reset(0));
        actionSubject.next(countersActions.addCounter());
        actionSubject.next(countersActions.removeCounter(2));
        expect(saveCountersStub.called).toBe(false);
        clock.tick(1000);
      })
      .then(() => {
        expect(output).toEqual([]);
        expect(saveCountersStub.calledOnceWithExactly(list)).toBe(true);
      });
    });

    it('should dispatch CountersError if saveCounters rejects', () => {
      // replace the standard stub with one that rejects
      saveCountersStub.restore();
      saveCountersStub = createSaveCountersStub(
        () => Promise.reject('error in saveCounters')
      );
      actionSubject.next(countersActions.addCounter());
      clock.tick(1000);
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(saveCountersStub.calledOnceWithExactly(list)).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(countersActions.countersUpdated([3,1,2]));
      actionSubject.next(countersActions.countersLoaded());
      actionSubject.next(countersActions.loadCounters());
      actionSubject.next(countersActions.countersError(new Error('an error')));
      clock.tick(1000);
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(saveCountersStub.called).toBe(false);
      });

    });

  });

  describe('loadUsersEpic', () => {

    const list = [
      {
        firstname: 'abc',
        id: '123',
        lastname: 'def'
      },
      {
        firstname: 'ghi',
        id: '456',
        lastname: 'jkl'
      }
    ];
    let loadUsersStub: sinon.SinonStub;
    const createLoadUsersStub = (fake: () => void) => 
      sinon.stub(usersApi, 'loadUsers').callsFake(fake)
    
    beforeEach(() => {
      loadUsersStub = createLoadUsersStub(
        () => Promise.resolve(list)
      );
      loadUsersEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      loadUsersStub.restore();
    });

    it('should dispatch UsersUpdated and UsersLoaded on LoadUsers if loadUsers succeeds', () => {
      actionSubject.next(usersActions.loadUsers());
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(loadUsersStub.calledOnce).toBe(true);
      });
    });

    it('should dispatch UsersError if loadUsers rejects', () => {
      loadUsersStub.restore();
      loadUsersStub = createLoadUsersStub(
        () => Promise.reject('error in loadUsers')
      );
      actionSubject.next(usersActions.loadUsers());
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(loadUsersStub.calledOnce).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(usersActions.deleteUser('1'));
      actionSubject.next(usersActions.usersError(new Error('an error')));
      actionSubject.next(usersActions.usersLoaded());
      actionSubject.next(usersActions.usersUpdated([]));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(loadUsersStub.called).toBe(false);
      });
    });

  });

  describe('createUserEpic', () => {

    const user = {
      firstname: 'abc',
      id: '123',
      lastname: 'def'
    }
    let createUserStub: sinon.SinonStub;
    const createCreateUserStub = (fake: () => void) => 
      sinon.stub(usersApi, 'createUser').callsFake(fake)
    
    beforeEach(() => {
      createUserStub = createCreateUserStub(
        () => Promise.resolve('987')
      );
      createUserEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      createUserStub.restore();
    });

    it('should call createUser on CreateUser', () => {
      actionSubject.next(usersActions.createUser(user));
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(createUserStub.calledOnceWithExactly(user)).toBe(true);
      });
    });

    it('should dispatch UsersError if createUser rejects', () => {
      createUserStub.restore();
      createUserStub = createCreateUserStub(
        () => Promise.reject('error in createUser')
      );
      actionSubject.next(usersActions.createUser(user));
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(createUserStub.calledOnceWithExactly(user)).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(usersActions.deleteUser('1'));
      actionSubject.next(usersActions.usersError(new Error('an error')));
      actionSubject.next(usersActions.usersLoaded());
      actionSubject.next(usersActions.usersUpdated([]));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(createUserStub.called).toBe(false);
      });
    });

  });

  describe('updateUserEpic', () => {

    const user = {
      firstname: 'abc',
      id: '123',
      lastname: 'def'
    }
    let updateUserStub: sinon.SinonStub;
    const createUpdateUserStub = (fake: () => void) => 
      sinon.stub(usersApi, 'updateUser').callsFake(fake)
    
    beforeEach(() => {
      updateUserStub = createUpdateUserStub(
        () => Promise.resolve()
      );
      updateUserEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      updateUserStub.restore();
    });

    it('should call updateUser on UpdateUser', () => {
      actionSubject.next(usersActions.updateUser(user));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(updateUserStub.calledOnceWithExactly(user)).toBe(true);
      });
    });

    it('should dispatch UsersError if updateUser rejects', () => {
      updateUserStub.restore();
      updateUserStub = createUpdateUserStub(
        () => Promise.reject('error in updateUser')
      );
      actionSubject.next(usersActions.updateUser(user));
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(updateUserStub.calledOnceWithExactly(user)).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(usersActions.deleteUser('1'));
      actionSubject.next(usersActions.usersError(new Error('an error')));
      actionSubject.next(usersActions.usersLoaded());
      actionSubject.next(usersActions.usersUpdated([]));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(updateUserStub.called).toBe(false);
      });
    });

  });

  describe('deleteUserEpic', () => {

    const id = '123';
    let deleteUserStub: sinon.SinonStub;
    const createDeleteUserStub = (fake: () => void) => 
      sinon.stub(usersApi, 'deleteUser').callsFake(fake)
    
    beforeEach(() => {
      deleteUserStub = createDeleteUserStub(
        () => Promise.resolve()
      );
      deleteUserEpic(action$, state$, undefined).subscribe(
        a => output.push(a)
      );
    });

    afterEach(() => {
      deleteUserStub.restore();
    });

    it('should call deleteUser on DeleteUser', () => {
      actionSubject.next(usersActions.deleteUser(id));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(deleteUserStub.calledOnceWithExactly(id)).toBe(true);
      });
    });

    it('should dispatch UsersError if deleteUser rejects', () => {
      deleteUserStub.restore();
      deleteUserStub = createDeleteUserStub(
        () => Promise.reject('error in deleteUser')
      );
      actionSubject.next(usersActions.deleteUser(id));
      return Promise.resolve().then(() => {
        expect(output).toMatchSnapshot();
        expect(deleteUserStub.calledOnceWithExactly(id)).toBe(true);
      });
    });
  
    it('should do nothing on other actions', () => {
      actionSubject.next(usersActions.usersError(new Error('an error')));
      actionSubject.next(usersActions.usersLoaded());
      actionSubject.next(usersActions.usersUpdated([]));
      return Promise.resolve().then(() => {
        expect(output).toEqual([]);
        expect(deleteUserStub.called).toBe(false);
      });
    });

  });

});