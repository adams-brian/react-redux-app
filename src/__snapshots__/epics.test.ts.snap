// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`epics createUserEpic should dispatch UsersError if createUser rejects 1`] = `
Array [
  Object {
    "payload": "error in createUser",
    "type": "[Users] USERS ERROR",
  },
]
`;

exports[`epics deleteUserEpic should dispatch UsersError if deleteUser rejects 1`] = `
Array [
  Object {
    "payload": "error in deleteUser",
    "type": "[Users] USERS ERROR",
  },
]
`;

exports[`epics loadCountersEpics should dispatch CountersError if loadCounters rejects 1`] = `
Array [
  Object {
    "payload": "error in loadCounters",
    "type": "[Counters] COUNTERS ERROR",
  },
]
`;

exports[`epics loadCountersEpics should dispatch CountersUpdated and CountersLoaded on LoadCounters if loadCounters succeeds 1`] = `
Array [
  Object {
    "payload": Array [
      5,
      1,
      4,
      2,
      3,
    ],
    "type": "[Counters] COUNTERS UPDATED",
  },
  Object {
    "type": "[Counters] COUNTERS LOADED",
  },
]
`;

exports[`epics loadUsersEpic should dispatch UsersError if loadUsers rejects 1`] = `
Array [
  Object {
    "payload": "error in loadUsers",
    "type": "[Users] USERS ERROR",
  },
]
`;

exports[`epics loadUsersEpic should dispatch UsersUpdated and UsersLoaded on LoadUsers if loadUsers succeeds 1`] = `
Array [
  Object {
    "payload": Array [
      Object {
        "_id": "123",
        "firstname": "abc",
        "lastname": "def",
      },
      Object {
        "_id": "456",
        "firstname": "ghi",
        "lastname": "jkl",
      },
    ],
    "type": "[Users] USERS UPDATED",
  },
  Object {
    "type": "[Users] USERS LOADED",
  },
]
`;

exports[`epics saveCountersEpic should dispatch CountersError if saveCounters rejects 1`] = `
Array [
  Object {
    "payload": "error in saveCounters",
    "type": "[Counters] COUNTERS ERROR",
  },
]
`;

exports[`epics updateUserEpic should dispatch UsersError if updateUser rejects 1`] = `
Array [
  Object {
    "payload": "error in updateUser",
    "type": "[Users] USERS ERROR",
  },
]
`;
